File: vision/triangulation.py

import math

# Assuming all variables are defined and initialized
x1, y1, x2, y2, theta1, theta2 = 2, 0, 0, 0, 240, 120 # Replace with actual values

# Convert angles from degrees to radians
theta1_rad = math.radians(theta1)
theta2_rad = math.radians(theta2)

# Calculate x3 and y3
x3 = ((y1 - y2) + x2 * math.tan(theta2_rad) - x1 * math.tan(theta1_rad)) / (math.tan(theta2_rad) - math.tan(theta1_rad))

y3 = ((y1 * math.tan(theta2_rad) - y2 * math.tan(theta1_rad)) - ((x1 - x2) * math.tan(theta2_rad) * math.tan(theta1_rad))) / (math.tan(theta2_rad) - math.tan(theta1_rad))

print('x3:', x3)
print('y3:', y3)

File: vision/triangulation.js
function triangulate(beacon1, beacon2, angle1, angle2) {
    // convert angles to radians
    let angle1Rad = Math.PI / 180 * angle1;
    let angle2Rad = Math.PI / 180 * angle2;

    // find the coordinates of the current location
    let x_current = ((beacon1.y - beacon2.y) + beacon2.x * Math.tan(angle2Rad) - beacon1.x * Math.tan(angle1Rad)) / (Math.tan(angle2Rad) - Math.tan(angle1Rad));
    let y_current = ((beacon1.y * Math.tan(angle2Rad) - beacon2.y * Math.tan(angle1Rad)) - ((beacon1.x - beacon2.x) * Math.tan(angle2Rad) * Math.tan(angle1Rad))) / (Math.tan(angle2Rad) - Math.tan(angle1Rad));

    return {x: x_current, y: y_current};
}

// Testing: (Works!)
// Example coordinates of beacons:
let beacon1 = {x: 0, y: 0};
let beacon2 = {x: 5, y: 0};

let angle1 = 60;  // angle in degrees
let angle2 = 120;  // angle in degrees

let position = triangulate(beacon1, beacon2, angle1, angle2);
console.log(position);  // {x: ..., y: ...}
